### PR TITLE
[SUP-498] Pause cloud compute from clusters page

### PR DESCRIPTION
**Pull request by kyuksel**
### For Reviewers
I'm experimenting with something new in hopes of easing PR reviewing (feedback welcome) where the PR can be viewed in three separate diff's as listed below. 
1. [Functional changes](https://github.com/DataBiosphere/terra-ui/pull/2611/files/807da8446e5bd9cb6b14dd933b502f15e4fca749)
2. [Mostly refactoring/non-functional changes](https://github.com/DataBiosphere/terra-ui/pull/2611/files/807da8446e5bd9cb6b14dd933b502f15e4fca749)
3. [IntelliJ reformatter changes](https://github.com/DataBiosphere/terra-ui/pull/2611/commits/9a1164b3bc6fa05abd8534525dc48d924dfda232)
4. [PR Feedback #1](https://github.com/DataBiosphere/terra-ui/pull/2611/files/9a1164b3bc6fa05abd8534525dc48d924dfda232..a33c112a67b6f13be7ba71dc27d892750ee70103) to
  i. improve responsiveness of the page after clicking `Pause`
  ii. fix a pre-existing case where a cell would be displayed as blank upon clicking `Delete` for a disk while it's in `Deleting` status
  iii. make minor non-functional  improvements

### Ticket
https://broadworkbench.atlassian.net/browse/SUP-498

#### Dev
image

#### PR Branch
Screen Shot 2021-09-03 at 2 22 48 PM

Tooltip message when buttons are disabled:
image

### Highlights
-  Name the last column on the tables on `Cloud Environments` page as `Actions` and `Action`.
-  Add `Pause` button alongside `Delete` button on the `Cloud Environments` table.
-  Label `Pause` and `Delete` buttons.
-  Display disabled buttons in a lighter shade of grey for higher contrast with enabled buttons.
-  Do some refactoring along the way to reduce code duplications.

### Testing
- Locally deployed the changes and ran against dev backends.

### CODE REVIEW

**Comment by rtitle** [comment]
I tested a little bit, a few comments:

1. I feel like pause from the Cloud Environments page suffers from the "immediate feedback" problem. When I click pause nothing happens initially -- only after the page auto-refreshes ~10 seconds later do I see the status change to "Stopping". It feels like it'd be nicer to have the change reflected on the page immediately after click.

2. When a runtime or Galaxy is paused, I see a greyed-out Pause button. Should there be a Resume button instead?

3. I see a browser console warning when clicking pause:
```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at ComputeModalBase (http://localhost:3000/static/js/1.chunk.js:3510:25)
```

**Review by rtitle — CHANGES_REQUESTED** [review]
see comment above

File: src/components/runtime-common.js [hunk 1]
```diff
@@ -110,7 +110,7 @@ export const PeriodicCookieSetter = () => {
   return null
 }
 
-export const SaveFilesHelp = () => {
+export const SaveFilesHelp = (isGalaxyDisk = false) => {
```
Comment (kyuksel) [inline src/components/runtime-common.js:113]: This wasn't part of my ticket but while testing my changes, I noticed that we are displaying a piece of message specific to Jupyter notebooks (from line#122 below) when Galaxy disk is to be deleted, and wanted to fix that as well.

File: src/pages/Environments.js [hunk 2]
```diff
@@ -241,22 +250,33 @@ const Environments = ({ namespace }) => {
     }, [h(Link, ['details'])])
   }
```
Comment (kyuksel) [inline src/pages/Environments.js:250]: `renderDeleteButton` replaces `renderDeleteButtonApps` and `renderDeleteButtonRuntimes` to reduce code duplication.

File: src/pages/Environments.js [hunk 3]
```diff
@@ -108,7 +108,7 @@ const DeleteAppModal = ({ app: { googleProject, appName, diskName }, onDismiss,
   ])
 }
 
-const Environments = ({ namespace }) => {
```
Comment (kyuksel) [inline src/pages/Environments.js:111]: `namespace` wasn't being used

**Comment by kyuksel** [comment]
Thanks for the testing @rtitle.

> 1. I feel like pause from the Cloud Environments page suffers from the "immediate feedback" problem. When I click pause nothing happens initially -- only after the page auto-refreshes ~10 seconds later do I see the status change to "Stopping". It feels like it'd be nicer to have the change reflected on the page immediately after click.

I agree. This exists in `dev` too but I'll see if I can improve it opportunistically.

> 2. When a runtime or Galaxy is paused, I see a greyed-out Pause button. Should there be a Resume button instead?

No, this ticket was scoped to only having the `pause` functionality. I agree though that it'd be nice to add a `resume` functionality at which time it'd make sense to have the button label and icons toggle between `pause` and `resume`.

> 3. I see a browser console warning when clicking pause

TL;DR: It's been a known situation and doesn't indicate a problem in our case. 

We have a wrapper around API calls that allows for setting a flag before and after making the API call. The one that's made after the API call triggers this warning because it happens after the component is unmounted. Since the flag we're updating isn't dependent on the component being mounted, it doesn't cause any issues for us.

**Comment by kyuksel** [comment]
> I feel like pause from the Cloud Environments page suffers from the "immediate feedback" problem. When I click pause nothing happens initially -- only after the page auto-refreshes ~10 seconds later do I see the status change to "Stopping". It feels like it'd be nicer to have the change reflected on the page immediately after click.

@rtitle This issue should be resolved now. Upon clicking `Pause`, the UI will display the usual spinner while it gets a response from Leo. Upon receiving the response, the spinner will disappear and the tables will have updated values/button states.

File: src/libs/runtime-utils.js [hunk 4]
```diff
@@ -226,10 +225,26 @@ export const getGalaxyCostTextChildren = (app, galaxyDataDisks) => {
...
 }
 
-export const isAppDeletable = app => _.includes(app?.status, ['RUNNING', 'ERROR'])
+/**
+ * 'Deletable' and 'Pausable' statuses are defined in a resource's respective Leonardo model:
+ * https://github.com/DataBiosphere/leonardo/blob/3339ae218b4258f704702475be1431b48a5e2932/core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala
+ * https://github.com/DataBiosphere/leonardo/blob/706a7504420ea4bec686d4f761455e8502b2ddf1/core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala
+ * https://github.com/DataBiosphere/leonardo/blob/e60c71a9e78b53196c2848cd22a752e22a2cf6f5/core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
+ */
+const deletableStatuses = resourceType => Utils.switchCase(resourceType,
```
Comment (zarsky-broad) [inline src/libs/runtime-utils.js]:
```js
export const isResourceDeletable = _.curry((resourceType, resource) => _.includes(_.capitalize(resource?.status),
  Utils.switchCase(resourceType,
    ['runtime', () => ['Unknown', 'Running', 'Updating', 'Error', 'Stopping', 'Stopped', 'Starting']],
    ['app', () => ['Unspecified', 'Running', 'Error']],
    ['disk', () => ['Failed', 'Ready']],
    [Utils.DEFAULT, () => console.error(`Resource type ${resourceType} is not one of runtime, app, or disk; deletability unknown`)]
  )
))

export const isComputePausable = _.curry((computeType, compute) => _.includes(_.capitalize(compute?.status),
  Utils.switchCase(computeType,
    ['runtime', () => ['Unknown', 'Running', 'Updating', 'Starting']],
    ['app', () => ['Running', 'Starting']],
    [Utils.DEFAULT, () => console.error(`Compute type ${computeType} is not one of runtime or app; pausability unknown`)]
  )
))
```

File: src/pages/Environments.js [hunk 5]
```diff
@@ -15,9 +15,13 @@ import { Ajax } from 'src/libs/ajax'
...
 import { withErrorIgnoring, withErrorReporting } from 'src/libs/error'
-import { getCurrentApp, getCurrentRuntime, getGalaxyComputeCost, getGalaxyCost, persistentDiskCostMonthly, runtimeCost } from 'src/libs/runtime-utils'
+import {
+  getCurrentApp, getCurrentRuntime, getGalaxyComputeCost, getGalaxyCost, isComputePausable, isResourceDeletable, persistentDiskCostMonthly,
+  runtimeCost
+} from 'src/libs/runtime-utils'
 import * as Style from 'src/libs/style'
-import { cond, formatUSD, makeCompleteDate, switchCase, useCancellation, useGetter, useOnMount, usePollingEffect, withBusyState } from 'src/libs/utils'
+import * as Utils from 'src/libs/utils'
+import { cond, formatUSD, makeCompleteDate, useCancellation, useGetter, useOnMount, usePollingEffect, withBusyState } from 'src/libs/utils'
```
Comment (zarsky-broad) [inline src/pages/Environments.js]: Please call utils as `Utils.foo`

File: src/pages/Environments.js [hunk 6]
```diff
@@ -126,6 +126,8 @@ const Environments = ({ namespace }) => {
   const [sort, setSort] = useState({ field: 'project', direction: 'asc' })
   const [diskSort, setDiskSort] = useState({ field: 'project', direction: 'asc' })
 
+  const disabledButtonColorDarkness = colors.dark(0.4)
```
Comment (zarsky-broad) [inline src/pages/Environments.js]: 0.72 passes contrast; change default plz, file tix with ux

File: src/pages/Environments.js [hunk 7]
```diff
@@ -158,6 +160,14 @@ const Environments = ({ namespace }) => {
 
   const loadData = withErrorReporting('Error loading cloud environments', refreshData)
 
+  const pauseComputeAndRefresh = withErrorReporting('Error loading cloud environments', async (computeType, compute) => {
+    const wrappedPauseCompute = withBusyState(setLoading, () => computeType === 'runtime' ?
+      Ajax().Runtimes.runtime(compute.googleProject, compute.runtimeName).stop() :
+      Ajax().Apps.app(compute.googleProject, compute.appName).pause())
+    await wrappedPauseCompute()
+    refreshData()
+  })
```
Comment (zarsky-broad) [inline src/pages/Environments.js]:
```suggestion
  const pauseComputeAndRefresh = withBusyState(setLoading, async (computeType, compute) => {
    const wrappedPauseCompute = withErrorReporting('Error stopping compute', () => computeType === 'runtime' ?
      Ajax().Runtimes.runtime(compute.googleProject, compute.runtimeName).stop() :
      Ajax().Apps.app(compute.googleProject, compute.appName).pause())
    await wrappedPauseCompute()
    await loadData()
  })
```
but also wrap error more specifically for stop/pause

File: src/pages/Environments.js [hunk 8]
```diff
@@ -366,20 +389,23 @@ const Environments = ({ namespace }) => {
...
               const cloudEnvironment = filteredCloudEnvironments[rowIndex]
               return cloudEnvironment.appName ? formatUSD(getGalaxyComputeCost(cloudEnvironment)) : formatUSD(runtimeCost(cloudEnvironment))
             }
           },
           {
-            size: { basis: 50, grow: 0 },
-            headerRenderer: () => div({ className: 'sr-only' }, ['Actions']),
+            size: { basis: 200, grow: 0 },
+            headerRenderer: () => 'Actions',
+            field: 'actions',
```
Comment (zarsky-broad) [inline src/pages/Environments.js]:
```suggestion
```

File: src/pages/Environments.js [hunk 9]
```diff
@@ -366,20 +389,23 @@ const Environments = ({ namespace }) => {
...
             cellRenderer: ({ rowIndex }) => {
               const cloudEnvironment = filteredCloudEnvironments[rowIndex]
-              return cloudEnvironment.appName ? renderDeleteButtonApps(cloudEnvironment) : renderDeleteButtonRuntimes(cloudEnvironment)
+              const computeType = cloudEnvironment.appName ? 'app' : 'runtime'
```
Comment (zarsky-broad) [inline src/pages/Environments.js]:
```suggestion
              const computeType = !!cloudEnvironment.appName ? 'app' : 'runtime'
```

File: src/pages/Environments.js [hunk 10]
```diff
@@ -477,19 +506,23 @@ const Environments = ({ namespace }) => {
...
-            headerRenderer: () => div({ className: 'sr-only' }, ['Actions']),
+            size: { basis: 200, grow: 0 },
+            field: 'action',
+            headerRenderer: () => 'Action',
             cellRenderer: ({ rowIndex }) => {
               const { id, status, name } = filteredDisks[rowIndex]
               const error = cond(
                 [status === 'Creating', () => 'Cannot delete this disk because it is still being created'],
-                [_.some({ runtimeConfig: { persistentDiskId: id } }, runtimes) || _.some({ diskName: name }, apps), () => 'Cannot delete this disk because it is attached. You must delete the cloud environment first.']
+                [status === 'Deleting', () => 'The disk is being deleted'],
```
Comment (zarsky-broad) [inline src/pages/Environments.js]:
```suggestion
                [status === 'Deleting', () => 'The disk is being deleted.'],
```

**Review by zarsky-broad — APPROVED** [review]
Pending comments 👍

**Review by rtitle — APPROVED** [review]
Retested, looks good to me. Thanks for addressing my comments!
